require-trusted-types-for CSP parsing incorrect

lukewarlow opened the pull request:
#### 18919c9e6b223759f65a2a2ddbc5d65e6bb2e611
<pre>
require-trusted-types-for CSP parsing incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=289904">https://bugs.webkit.org/show_bug.cgi?id=289904</a>

Reviewed by Anne van Kesteren.

This patch updates the parsing to ensure that &apos;script&apos; is only valid when it&apos;s followed by a whitespace
or end of buffer.

It also updates the parsing to no longer early return when it finds an invalid sink group.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/should-sink-type-mismatch-violation-be-blocked-by-csp-001-expected.txt: Added.
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::ContentSecurityPolicyDirectiveList::parseRequireTrustedTypesFor):

Canonical link: <a href="https://commits.webkit.org/300770@main">https://commits.webkit.org/300770@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1659a1575b4b2c9cdbd2b6c76c231b8a366cb23d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123761 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43476 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34173 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130539 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75911 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125638 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44200 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52070 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94124 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62463 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126714 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35215 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110718 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74727 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34175 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28877 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74019 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104939 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29101 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133229 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50713 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38621 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102601 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51088 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106937 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102430 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26045 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47770 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26023 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47558 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50567 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56331 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50041 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53387 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51715 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e11b605cea41808bdafa510a0e72e9594c49f229. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43118

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/should-sink-type-mismatch-violation-be-blocked-by-csp-001.html:
```diff
@@ -113,7 +113,7 @@
     let results = await trySendingPlainStringToTrustedTypeSink(
       ["script"],
-      "header(Content-Security-Policy,require-trusted-types-for 'script''script' 'script',True)",
+      "header(Content-Security-Policy,require-trusted-types-for 'script''script',True)",
```
The existing test is wrong this corrects it. This change will be exported separately by an associated Gecko patch.

annevk commented on Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
```diff
@@ -622,11 +622,12 @@ void ContentSecurityPolicyDirectiveList::parseRequireTrustedTypesFor(ParsedDirec
 
             auto begin = buffer.position();
-            if (skipExactlyIgnoringASCIICase(buffer, "'script'"_s))
+            if (skipExactlyIgnoringASCIICase(buffer, "'script'"_s) && (buffer.atEnd() || isUnicodeCompatibleASCIIWhitespace(*buffer)))
```
This is the wrong kind of whitespace. Though maybe we should fix that as part of #13197. Would it be easy to add whitespace tests?

annevk commented on Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
```diff
@@ -622,11 +622,12 @@ void ContentSecurityPolicyDirectiveList::parseRequireTrustedTypesFor(ParsedDirec
             else {
-                policy().reportInvalidTrustedTypesSinkGroup(std::span { begin, buffer.position() });
-                return;
+                skipWhile<isNotASCIISpace>(buffer);
```
Ditto.

annevk commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/should-sink-type-mismatch-violation-be-blocked-by-csp-001-expected.txt:
```diff
@@ -0,0 +1,68 @@
+PASS Multiple report-only require-trusted-types-for directives.
+FAIL One violated report-only require-trusted-types-for directive followed by multiple enforce directives assert_equals: expected 5 but got 1
+FAIL One violated enforce require-trusted-types-for directive followed by multiple report-only directives assert_equals: expected 5 but got 1
+FAIL Mixing enforce and report-only require-trusted-types-for directives. assert_equals: expected 6 but got 1
```
Do we have some kind of follow-up filed for these failures?

lukewarlow replied:
Not yet, I can do though. I'm not entirely sure if they are test issues on WebKit ones.

lukewarlow replied:
Typically, I've just found what I think is the cause, I'll make a follow up patch to fix that. Per the spec the test is correct.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1659a1575b4b2c9cdbd2b6c76c231b8a366cb23d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43118

lukewarlow committed:
require-trusted-types-for CSP parsing incorrect
https://bugs.webkit.org/show_bug.cgi?id=289904

Reviewed by Anne van Kesteren.

This patch updates the parsing to ensure that 'script' is only valid when it's followed by a whitespace
or end of buffer.

It also updates the parsing to no longer early return when it finds an invalid sink group.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/should-sink-type-mismatch-violation-be-blocked-by-csp-001-expected.txt: Added.
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::ContentSecurityPolicyDirectiveList::parseRequireTrustedTypesFor):

Canonical link: https://commits.webkit.org/300770@main
